growthbook.go

```diff
@@ -97,136 +346,277 @@ func (fr *FeatureResult) GetValueWithDefault(def FeatureValue) FeatureValue {
+	gb.inner.subscriptions[id] = callback
+	gb.inner.nextSubscriptionID++
+	return func() {
+		delete(gb.inner.subscriptions, id)
+	}
+}
+
+// GetAllResults returns a map containing all the latest results from
+// all experiments that have been run, indexed by the experiment key.
+func (gb *GrowthBook) GetAllResults() map[string]*Assignment {
+	gb.inner.RLock()
+	defer gb.inner.RUnlock()
+
+	return gb.inner.assigned
+}
+
+// ClearSavedResults clears out any experiment results saved within a
+// GrowthBook instance (used for deciding whether to send data to
+// subscriptions).
+func (gb *GrowthBook) ClearSavedResults() {
+	gb.inner.Lock()
+	defer gb.inner.Unlock()
+
+	gb.inner.assigned = make(map[string]*Assignment)
+}
+
+// ClearTrackingData clears out records of calls to the experiment
+// tracking callback.
+func (gb *GrowthBook) ClearTrackingData() {
+	gb.inner.Lock()
+	defer gb.inner.Unlock()
+
+	gb.inner.trackedExperiments = make(map[string]bool)
+}
+
+// GetAPIInfo gets the hostname and client key for GrowthBook API
+// access.
+func (gb *GrowthBook) GetAPIInfo() (string, string) {
+	gb.inner.RLock()
+	defer gb.inner.RUnlock()
+
+	apiHost := gb.inner.context.APIHost
+	if apiHost == "" {
+		apiHost = "https://cdn.growthbook.io"
+	}
+
+	return strings.TrimRight(apiHost, "/"), gb.inner.context.ClientKey
+}
+
+type FeatureRepoOptions struct {
+	AutoRefresh bool
+	Timeout     time.Duration
+	SkipCache   bool
+}
+
+func (gb *GrowthBook) LoadFeatures(options *FeatureRepoOptions) {
+	gb.refresh(options, true, true)
+	if options != nil && options.AutoRefresh {
+		RepoSubscribe(gb)
+	}
+}
+
+func (gb *GrowthBook) RefreshFeatures(options *FeatureRepoOptions) {
+	gb.refresh(options, false, true)
+}
+
+func (gb *GrowthBook) refresh(
+	options *FeatureRepoOptions, allowStale bool, updateInstance bool) {
+
+	if gb.inner.context.ClientKey == "" {
+		logError("Missing clientKey")
+		return
+	}
+	var timeout time.Duration
+	skipCache := gb.inner.context.DevMode
+	if options != nil {
+		timeout = options.Timeout
+		skipCache = skipCache || options.SkipCache
+	}
+	ConfigureCacheStaleTTL(gb.inner.context.CacheTTL)
+	RepoRefreshFeatures(gb, timeout, skipCache, allowStale, updateInstance)
+}
+
+//-- PRIVATE FUNCTIONS START HERE ----------------------------------------------
+
+func (gb *GrowthBook) trackFeatureUsage(key string, res *FeatureResult) {
+	// Don't track feature usage that was forced via an override.
+	if res.Source == OverrideResultSource {
+		return
 	}
 
-	// Fall back to using the default value
-	return getFeatureResult(feature.DefaultValue, DefaultValueResultSource, nil, nil)
+	// Only track a feature once, unless the assigned value changed.
+	if saved, ok := gb.inner.trackedFeatures[key]; ok && reflect.DeepEqual(saved, res.Value) {
+		return
+	}
+	gb.inner.trackedFeatures[key] = res.Value
+
+	// Fire user-supplied callback
+	if gb.inner.context.OnFeatureUsage != nil {
+		gb.inner.context.OnFeatureUsage(key, res)
+	}
```

Comment:
Nice. This will be the only SDK other than the JS ones that has this support, so that's good.

Reply:
Yeah, Graham specifically mentioned it as a thing that some people had been asking for!